messaging/schema.py

```diff
@@ -0,0 +1,243 @@
+
+
+class MessageTranslationType(DjangoObjectType):
+    class Meta:
+        model = MessageTranslation
+        exclude = ("id", "master")
+
+
+class RecipientSelectionEnum(graphene.Enum):
+    ALL = "all"
+    INVITED = "invited"
+    ENROLLED = "enrolled"
+    ATTENDED = "attended"
+    SUBSCRIBED_TO_FREE_SPOT_MESSAGE = "subscribed_to_free_spot_message"
+
+
+class MessageNode(DjangoObjectType):
+    subject = graphene.String()
+    body_text = graphene.String()
+    recipient_selection = graphene.Field(RecipientSelectionEnum)
+
+    class Meta:
+        model = Message
+        interfaces = (relay.Node,)
+        fields = (
+            "id",
+            "project",
+            "created_at",
+            "updated_at",
+            "subject",
+            "body_text",
+            "recipient_selection",
+            "event",
+            "occurrences",
+            "sent_at",
+            "recipient_count",
+            "translations",
+        )
+
+    @classmethod
+    def get_queryset(cls, queryset, info):
+        return queryset.prefetch_related("translations")
+
+
+class MessageTranslationsInput(graphene.InputObjectType):
+    language_code = LanguageEnum(required=True)
+    subject = graphene.String()
+    body_text = graphene.String()
+
+
+class AddMessageMutation(graphene.relay.ClientIDMutation):
+    class Input:
+        translations = graphene.List(MessageTranslationsInput)
+        project_id = graphene.GlobalID()
+        recipient_selection = RecipientSelectionEnum(required=True)
+        event_id = graphene.ID()
+        occurrence_ids = graphene.List(graphene.NonNull(graphene.ID))
+
+    message = graphene.Field(MessageNode)
+
+    @classmethod
+    @project_user_required
+    @transaction.atomic
+    def mutate_and_get_payload(cls, root, info, **kwargs):
+        data = deepcopy(kwargs)
+
+        data["project"] = get_obj_if_user_can_administer(
+            info, data.pop("project_id"), Project
+        )
+
+        if "event_id" in data:
+            data["event"] = get_obj_if_user_can_administer(
+                info, data.pop("event_id"), Event
+            )
+        else:
+            data["event"] = None
+        occurrences = [
+            get_obj_if_user_can_administer(info, occurrence_id, Occurrence)
+            for occurrence_id in data.pop("occurrence_ids", [])
+        ]
+
+        validate_event_and_occurrences(data["event"], occurrences)
+
+        message = Message.objects.create_translatable_object(**data)
+        if occurrences:
+            message.occurrences.set(occurrences)
+
+        logger.info(
+            f"user {info.context.user.uuid} added message {message} with data {kwargs}"
+        )
+
+        return AddMessageMutation(message=message)
+
+
+class UpdateMessageMutation(graphene.relay.ClientIDMutation):
+    class Input:
+        id = graphene.GlobalID()
+        translations = graphene.List(MessageTranslationsInput)
+        project_id = graphene.GlobalID(required=False)
```

Comment:
Could be simply `ID()` then

Reply:
Yeah I guess it could, will change. I don't really get that graphene `GlobalID` vs `ID` anyway.